File: ts/src/verifier.test.ts
import { test } from "node:test";
import { strict as assert } from "node:assert";
import { Verifier } from "./verifier.js";
import { RealmError } from "./errors.js";

const BASE_URL = "https://auth.test.example";
const REALM_ID = "01HXYZREALM";
const AUD = "example.com";

/** Generate RSA keypair, export JWK for public, return signer fn + jwk. */
async function mintKey(kid: string) {
  const kp = await globalThis.crypto.subtle.generateKey(
    {
      name: "RSASSA-PKCS1-v1_5",
      modulusLength: 2048,
      publicExponent: new Uint8Array([1, 0, 1]),
      hash: "SHA-256",
    },
    true,
    ["sign", "verify"],
  );
  const jwk = await globalThis.crypto.subtle.exportKey("jwk", kp.publicKey);
  const publicJwk = { kty: jwk.kty!, n: jwk.n!, e: jwk.e!, kid, alg: "RS256", use: "sig" };
  async function signToken(claims: Record<string, unknown>): Promise<string> {
    const header = { alg: "RS256", typ: "JWT", kid };
    const enc = (obj: unknown) => b64urlEncode(new TextEncoder().encode(JSON.stringify(obj)));
    const signingInput = `${enc(header)}.${enc(claims)}`;
    const sig = await globalThis.crypto.subtle.sign(
      { name: "RSASSA-PKCS1-v1_5" },
      kp.privateKey,
      new TextEncoder().encode(signingInput),
    );
    return `${signingInput}.${b64urlEncode(new Uint8Array(sig))}`;
  }
  return { publicJwk, signToken };
}

function b64urlEncode(bytes: Uint8Array): string {
  let bin = "";
  for (const b of bytes) bin += String.fromCharCode(b);
  return btoa(bin).replace(/=+$/, "").replace(/\+/g, "-").replace(/\//g, "_");
}

function fakeFetch(realmKeys: Record<string, object[]>): {
  fetch: typeof fetch;
  calls: { count: number };
} {
  const calls = { count: 0 };
  const fetch = (async (input: RequestInfo | URL): Promise<Response> => {
    calls.count++;
    const url = typeof input === "string" ? input : input.toString();
    for (const realm of Object.keys(realmKeys)) {
      if (url.endsWith(`/${realm}/.well-known/jwks.json`)) {
        return new Response(JSON.stringify({ keys: realmKeys[realm] }), {
          status: 200,
          headers: { "content-type": "application/json" },
        });
      }
    }
    return new Response("not found", { status: 404 });
  }) as typeof fetch;
  return { fetch, calls };
}

test("verify: happy path", async () => {
  const { publicJwk, signToken } = await mintKey("kid-1");
  const { fetch, calls } = fakeFetch({ [REALM_ID]: [publicJwk] });

  const now = new Date("2026-04-01T00:00:00Z");
  const v = new Verifier({ baseUrl: BASE_URL, audience: AUD, fetch, now: () => now });

  const exp = Math.floor(now.getTime() / 1000) + 900;
  const token = await signToken({
    iss: `${BASE_URL}/${REALM_ID}`,
    sub: "user-1",
    aud: AUD,
    exp,
    iat: Math.floor(now.getTime() / 1000),
    tenant_id: "tenant-1",
    role: "owner",
  });

  const claims = await v.verify(token);
  assert.equal(claims.sub, "user-1");
  assert.equal(claims.tenant_id, "tenant-1");
  assert.equal(calls.count, 1);

  await v.verify(token);
  assert.equal(calls.count, 1);
});

test("verify: wrong audience rejected", async () => {
  const { publicJwk, signToken } = await mintKey("kid-1");
  const { fetch } = fakeFetch({ [REALM_ID]: [publicJwk] });
  const now = new Date("2026-04-01T00:00:00Z");
  const v = new Verifier({ baseUrl: BASE_URL, audience: AUD, fetch, now: () => now });

  const token = await signToken({
    iss: `${BASE_URL}/${REALM_ID}`,
    sub: "u",
    aud: "wrong.example",
    exp: Math.floor(now.getTime() / 1000) + 900,
  });
  await assert.rejects(() => v.verify(token), (e: Error) => {
    return e instanceof RealmError && e.code === "wrong_audience";
  });
});

test("verify: wrong issuer rejected", async () => {
  const { publicJwk, signToken } = await mintKey("kid-1");
  const { fetch } = fakeFetch({ [REALM_ID]: [publicJwk] });
  const now = new Date("2026-04-01T00:00:00Z");
  const v = new Verifier({ baseUrl: BASE_URL, audience: AUD, fetch, now: () => now });

  const token = await signToken({
    iss: `https://evil.example/${REALM_ID}`,
    sub: "u",
    aud: AUD,
    exp: Math.floor(now.getTime() / 1000) + 900,
  });
  await assert.rejects(() => v.verify(token), (e: Error) => {
    return e instanceof RealmError && e.code === "wrong_issuer";
  });
});

test("verify: expired rejected", async () => {
  const { publicJwk, signToken } = await mintKey("kid-1");
  const { fetch } = fakeFetch({ [REALM_ID]: [publicJwk] });
  const now = new Date("2026-04-01T00:00:00Z");
  const v = new Verifier({ baseUrl: BASE_URL, audience: AUD, fetch, now: () => now, leewaySeconds: 0 });

  const token = await signToken({
    iss: `${BASE_URL}/${REALM_ID}`,
    sub: "u",
    aud: AUD,
    exp: Math.floor(now.getTime() / 1000) - 1,
  });
  await assert.rejects(() => v.verify(token), (e: Error) => {
    return e instanceof RealmError && e.code === "expired";
  });
});

test("verify: tampered signature rejected", async () => {
  const { publicJwk, signToken } = await mintKey("kid-1");
  const { fetch } = fakeFetch({ [REALM_ID]: [publicJwk] });
  const now = new Date("2026-04-01T00:00:00Z");
  const v = new Verifier({ baseUrl: BASE_URL, audience: AUD, fetch, now: () => now });

  const good = await signToken({
    iss: `${BASE_URL}/${REALM_ID}`,
    sub: "u",
    aud: AUD,
    exp: Math.floor(now.getTime() / 1000) + 900,
  });
  const parts = good.split(".");
  parts[2] = parts[2]!.slice(0, -1) + (parts[2]!.slice(-1) === "A" ? "B" : "A");
  const bad = parts.join(".");

  await assert.rejects(() => v.verify(bad), (e: Error) => {
    return e instanceof RealmError && e.code === "bad_signature";
  });
});

test("verify: unknown kid triggers refetch", async () => {
  const a = await mintKey("kid-old");
  const b = await mintKey("kid-new");

  let call = 0;
  const fetch = (async (input: RequestInfo | URL): Promise<Response> => {
    call++;
    const url = typeof input === "string" ? input : input.toString();
    if (!url.endsWith(`/${REALM_ID}/.well-known/jwks.json`)) {
      return new Response("not found", { status: 404 });
    }
    const keys = call === 1 ? [a.publicJwk] : [a.publicJwk, b.publicJwk];
    return new Response(JSON.stringify({ keys }), { status: 200 });
  }) as typeof fetch;

  const now = new Date("2026-04-01T00:00:00Z");
  const v = new Verifier({ baseUrl: BASE_URL, audience: AUD, fetch, now: () => now });

  const tokOld = await a.signToken({
    iss: `${BASE_URL}/${REALM_ID}`, sub: "u", aud: AUD,
    exp: Math.floor(now.getTime() / 1000) + 900,
  });
  await v.verify(tokOld);
  assert.equal(call, 1);

  const tokNew = await b.signToken({
    iss: `${BASE_URL}/${REALM_ID}`, sub: "u", aud: AUD,
    exp: Math.floor(now.getTime() / 1000) + 900,
  });
  await v.verify(tokNew);
  assert.equal(call, 2);
});

test("constructor: missing config rejected", () => {
  assert.throws(() => new Verifier({ baseUrl: "", audience: "x" }));
  assert.throws(() => new Verifier({ baseUrl: "http://x" }));
});

test("verify: malformed token rejected", async () => {
  const now = new Date("2026-04-01T00:00:00Z");
  const v = new Verifier({ baseUrl: BASE_URL, audience: AUD, fetch: fakeFetch({}).fetch, now: () => now });

  await assert.rejects(() => v.verify("not.a"), (e: Error) => {
    return e instanceof RealmError && e.code === "malformed";
  });
});

test("verify: audience auto-discovered via resolver and cached", async () => {
  const { publicJwk, signToken } = await mintKey("kid-1");
  const { fetch } = fakeFetch({ [REALM_ID]: [publicJwk] });
  const now = new Date("2026-04-01T00:00:00Z");
  let resolveCalls = 0;
  const v = new Verifier({
    baseUrl: BASE_URL,
    audienceResolver: async (rid) => {
      resolveCalls++;
      assert.equal(rid, REALM_ID);
      return AUD;
    },
    fetch,
    now: () => now,
  });

  const token = await signToken({
    iss: `${BASE_URL}/${REALM_ID}`, sub: "u", aud: AUD,
    exp: Math.floor(now.getTime() / 1000) + 900,
  });
  await v.verify(token);
  await v.verify(token);
  assert.equal(resolveCalls, 1, "resolver should be cached");
});

test("verify: per-call audience override wins over resolver", async () => {
  const { publicJwk, signToken } = await mintKey("kid-1");
  const { fetch } = fakeFetch({ [REALM_ID]: [publicJwk] });
  const now = new Date("2026-04-01T00:00:00Z");
  const v = new Verifier({
    baseUrl: BASE_URL,
    audienceResolver: async () => "would-be-wrong.example",
    fetch,
    now: () => now,
  });

  const token = await signToken({
    iss: `${BASE_URL}/${REALM_ID}`, sub: "u", aud: AUD,
    exp: Math.floor(now.getTime() / 1000) + 900,
  });
  const claims = await v.verify(token, { audience: AUD });
  assert.equal(claims.sub, "u");
});

File: go/verifier_test.go
package realmid

import (
	"context"
	"crypto"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"math/big"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
	"time"
)

const (
	testRealmID = "01HXYZREALM"
	testAud     = "example.com"
)

type signFn func(claims map[string]any) string

func mintKey(t *testing.T, kid string) (signFn, jwk) {
	t.Helper()
	priv, err := rsa.GenerateKey(rand.Reader, 2048)
	if err != nil {
		t.Fatalf("rsa generate: %v", err)
	}
	pub := priv.PublicKey
	publicJwk := jwk{
		Kty: "RSA",
		Kid: kid,
		Alg: "RS256",
		Use: "sig",
		N:   base64.RawURLEncoding.EncodeToString(pub.N.Bytes()),
		E:   base64.RawURLEncoding.EncodeToString(big.NewInt(int64(pub.E)).Bytes()),
	}
	sign := func(claims map[string]any) string {
		hdr := map[string]any{"alg": "RS256", "typ": "JWT", "kid": kid}
		hb, _ := json.Marshal(hdr)
		cb, _ := json.Marshal(claims)
		signing := base64.RawURLEncoding.EncodeToString(hb) + "." +
			base64.RawURLEncoding.EncodeToString(cb)
		sum := sha256.Sum256([]byte(signing))
		sig, err := rsa.SignPKCS1v15(rand.Reader, priv, crypto.SHA256, sum[:])
		if err != nil {
			t.Fatalf("sign: %v", err)
		}
		return signing + "." + base64.RawURLEncoding.EncodeToString(sig)
	}
	return sign, publicJwk
}

// fakeServer wires JWKS endpoints + a stub /platforms/mine + a stub
// /auth/login (platform_api_key grant) so a fully-configured *Realm can
// exercise the verifier end-to-end without the live API.
func fakeServer(t *testing.T, perRealm map[string][]jwk, audience string) *httptest.Server {
	t.Helper()
	mux := http.NewServeMux()
	for realm, keys := range perRealm {
		realm := realm
		keys := keys
		mux.HandleFunc("/"+realm+"/.well-known/jwks.json", func(w http.ResponseWriter, _ *http.Request) {
			_ = json.NewEncoder(w).Encode(jwksDoc{Keys: keys})
		})
	}
	mux.HandleFunc("/auth/login", func(w http.ResponseWriter, _ *http.Request) {
		_ = json.NewEncoder(w).Encode(map[string]any{
			"status":        "ok",
			"subject_type":  "platform",
			"refresh_token": "rtok-platform",
			"access_token":  "ptok_test_" + testRealmID,
			"expires_in":    300,
		})
	})
	mux.HandleFunc("/platforms/mine", func(w http.ResponseWriter, _ *http.Request) {
		_ = json.NewEncoder(w).Encode(map[string]any{
			"items": []any{
				map[string]any{
					"id":       testRealmID,
					"audience": audience,
					"domain":   audience,
				},
			},
		})
	})
	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		http.NotFound(w, r)
	})
	return httptest.NewServer(mux)
}

func baseClaims(srv string, opts ...func(map[string]any)) map[string]any {
	now := time.Now().Unix()
	c := map[string]any{
		"iss": srv + "/" + testRealmID,
		"sub": "01HUSER",
		"aud": testAud,
		"iat": now,
		"exp": now + 600,
	}
	for _, o := range opts {
		o(c)
	}
	return c
}

func mustRealm(t *testing.T, srv string) *Realm {
	t.Helper()
	r, err := NewRealm(Config{
		RealmID: testRealmID,
		APIKey:  "rk_live_test",
		BaseURL: srv,
	})
	if err != nil {
		t.Fatalf("NewRealm: %v", err)
	}
	return r
}

func TestVerify_HappyPath(t *testing.T) {
	sign, pub := mintKey(t, "kid-1")
	srv := fakeServer(t, map[string][]jwk{testRealmID: {pub}}, testAud)
	defer srv.Close()

	r := mustRealm(t, srv.URL)
	tok := sign(baseClaims(srv.URL, func(c map[string]any) {
		c["tenant_id"] = "01HTENANT"
		c["role"] = "owner"
		c["jti"] = "01HJTI"
		c["custom_field"] = "x"
	}))
	claims, err := r.Verify(context.Background(), tok, nil)
	if err != nil {
		t.Fatalf("verify: %v", err)
	}
	if claims.TenantID != "01HTENANT" {
		t.Errorf("tenant_id: got %q", claims.TenantID)
	}
	if claims.Role != "owner" {
		t.Errorf("role: got %q", claims.Role)
	}
	if claims.Extra["custom_field"] != "x" {
		t.Errorf("extra custom_field: got %v", claims.Extra["custom_field"])
	}
}

func TestVerify_Errors(t *testing.T) {
	sign, pub := mintKey(t, "kid-1")
	srv := fakeServer(t, map[string][]jwk{testRealmID: {pub}}, testAud)
	defer srv.Close()
	r := mustRealm(t, srv.URL)

	cases := []struct {
		name  string
		token string
		want  ErrorCode
	}{
		{"malformed-not-3-parts", "a.b", ErrCodeMalformed},
		{"wrong-aud", sign(baseClaims(srv.URL, func(c map[string]any) { c["aud"] = "other" })), ErrCodeWrongAudience},
		{"wrong-iss", sign(baseClaims(srv.URL, func(c map[string]any) { c["iss"] = "https://evil.example/" + testRealmID })), ErrCodeWrongIssuer},
		{"expired", sign(baseClaims(srv.URL, func(c map[string]any) { c["exp"] = time.Now().Add(-time.Hour).Unix() })), ErrCodeExpired},
		{"not-yet-valid", sign(baseClaims(srv.URL, func(c map[string]any) { c["nbf"] = time.Now().Add(time.Hour).Unix() })), ErrCodeNotYetValid},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			_, err := r.Verify(context.Background(), c.token, nil)
			if err == nil {
				t.Fatalf("expected error")
			}
			var verr *RealmError
			if !errors.As(err, &verr) {
				t.Fatalf("expected *RealmError, got %T", err)
			}
			if verr.Code != c.want {
				t.Errorf("code: got %s, want %s", verr.Code, c.want)
			}
		})
	}
}

func TestVerify_UnknownKID_RefetchesAndStillFails(t *testing.T) {
	_, pubKnown := mintKey(t, "kid-known")
	srv := fakeServer(t, map[string][]jwk{testRealmID: {pubKnown}}, testAud)
	defer srv.Close()
	r := mustRealm(t, srv.URL)

	signUnknown, _ := mintKey(t, "kid-unknown")
	tok := signUnknown(baseClaims(srv.URL))
	_, err := r.Verify(context.Background(), tok, nil)
	var verr *RealmError
	if !errors.As(err, &verr) || verr.Code != ErrCodeUnknownKID {
		t.Fatalf("expected unknown_kid, got %v", err)
	}
}

func TestVerify_BadSignature(t *testing.T) {
	sign, pub := mintKey(t, "kid-1")
	srv := fakeServer(t, map[string][]jwk{testRealmID: {pub}}, testAud)
	defer srv.Close()
	r := mustRealm(t, srv.URL)
	tok := sign(baseClaims(srv.URL))

	parts := strings.Split(tok, ".")
	mangled := parts[0] + "." + parts[1] + "." + base64.RawURLEncoding.EncodeToString([]byte("not-a-real-sig"))
	_, err := r.Verify(context.Background(), mangled, nil)
	var verr *RealmError
	if !errors.As(err, &verr) || verr.Code != ErrCodeBadSignature {
		t.Fatalf("expected bad_signature, got %v", err)
	}
}

func TestVerify_JWKSFetchFailed(t *testing.T) {
	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: "http://127.0.0.1:1"})
	sign, _ := mintKey(t, "kid-1")
	// Stub audience cache so the verifier does not block on info().
	r.info.cached = &RealmInfo{ID: testRealmID, Audience: testAud}
	tok := sign(map[string]any{
		"iss": "http://127.0.0.1:1/" + testRealmID,
		"sub": "x", "aud": testAud, "exp": time.Now().Unix() + 600,
	})
	_, err := r.Verify(context.Background(), tok, nil)
	var verr *RealmError
	if !errors.As(err, &verr) || verr.Code != ErrCodeJWKSFetchFailed {
		t.Fatalf("expected jwks_fetch_failed, got %v", err)
	}
}

func TestNewRealm_Validation(t *testing.T) {
	if _, err := NewRealm(Config{APIKey: "x"}); err == nil {
		t.Fatal("expected error for missing RealmID")
	}
	if _, err := NewRealm(Config{RealmID: "x"}); err == nil {
		t.Fatal("expected error for missing APIKey")
	}
}

// Compile-time check that fmt is used.
var _ = fmt.Sprintf
